src/core/execute.c

```diff
@@ -2103,20 +2102,17 @@ static int setup_exec_directory(
                         if (r < 0)
                                 goto fail;
 
-                        effective = pp;
-
+                        /* Lock down the access mode */
+                        if (chmod(pp, context->directories[type].mode) < 0) {
+                                r = -errno;
+                                goto fail;
+                        }
                 } else {
                         r = mkdir_label(p, context->directories[type].mode);
-                        if (r < 0 && r != -EEXIST)
+                        if (r == -EEXIST)
+                                return 0;
+                        if (r < 0)
                                 goto fail;
-
-                        effective = p;
-                }
-
-                /* First lock down the access mode */
-                if (chmod(effective, context->directories[type].mode) < 0) {
-                        r = -errno;
-                        goto fail;
```

Comment:
hmm, so why drop this part btw? if there's a mode configured i am pretty sure we should honour it and reapply it to the directory itself. After all, people might use this to set +s or so, which is entirely an OK thing to do, but would only work here if the directory already exists... specifiying the mode only during mkdir() is not a replacement, since tha means umask is in effect, and hence +s can never be set...

I mean, I am fine with dropping the chowning/chmodding down the tree if the dir already exists, but I am very sure we should apply the mode for the top-level item at least the way it was configured in the unit

Reply:
So... I don't think we can make a meaningful distinction between directory mode and ownership like this. They are closely entwined, and e.g. setting +s without chown would be bad. I think the only two realistic possibilities are to always keep existing directories as is, and I think this is preferable, or set all attributes on them, like was done previously. This is nicely explainable as "if the directory exists, it is left as is".